File: src/concurrent/java/Future.java
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   26 Mar 09  Brian Frank  Creation
//   24 Apr 20  Brian Frank  Make Future abstract
//
package fan.concurrent;

import fan.sys.*;
import java.util.concurrent.TimeUnit;
import java.util.ArrayList;

/**
 * Future is used to manage the entire lifecycle of each
 * message send to an Actor.  An actor's queue is a linked
 * list of messages.
 */
public abstract class Future
  extends FanObj
  implements java.util.concurrent.Future
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Future makeCompletable() { return new ActorFuture(null); }

  public static void make$(Future self, Future wraps)
  {
    self.wraps = wraps;
  }

  public Future() {}

//////////////////////////////////////////////////////////////////////////
// Obj
//////////////////////////////////////////////////////////////////////////

  public Type typeof() { return typeof$(); }

  public static Type typeof$()
  {
    if (type == null) type = Type.find("concurrent::Future");
    return type;
  }
  private static Type type;

//////////////////////////////////////////////////////////////////////////
// Future
//////////////////////////////////////////////////////////////////////////

  public final boolean isDone() { return status().isComplete(); }

  public final boolean isCancelled() { return status().isCancelled(); }

  public FutureStatus status()
  {
    return wrapped().status();
  }

  public Object get() { return get(null); }
  public final Object get(long t, TimeUnit u) { return get(Duration.make(u.toNanos(t))); }
  public Object get(Duration timeout)
  {
    return wrapped().get(timeout);
  }

  public Err err()
  {
    return wrapped().err();
  }

  public final Future waitFor() { return waitFor(null); }
  public Future waitFor(Duration timeout)
  {
    wrapped().waitFor(timeout);
    return this;
  }

  public final Future then(Func onOk) { return then(onOk, null); }
  public Future then(Func onOk, Func onErr)
  {
    return wrap(wrapped().then(onOk, onErr));
  }

  public static final void waitForAll(List<Future> list) { waitForAll(list, null); }
  public static final void waitForAll(List<Future> list, Duration timeout)
  {
    if (timeout == null)
    {
      for (int i=0; i<list.sz(); ++i)
      {
        Future f = (Future)list.get(i);
        f.waitFor(null);
      }
    }
    else
    {
      long deadline = Duration.nowMillis() + timeout.millis();
      for (int i=0; i<list.sz(); ++i)
      {
        Future f = (Future)list.get(i);
        long left = deadline - Duration.nowMillis();
        f.waitFor(Duration.makeMillis(left));
      }
    }
  }

  // java Future version
  public final boolean cancel(boolean mayInterrupt)
  {
    cancel();
    return true;
  }

  public void cancel()
  {
    wrapped().cancel();
  }

  public Future complete(Object r)
  {
    wrapped().complete(r);
    return this;
  }

  public Future completeErr(Err e)
  {
    wrapped().completeErr(e);
    return this;
  }

  public Object promise()
  {
    throw UnsupportedErr.make("Not available in Java VM");
  }

  public abstract Future wrap(Future wrap);

  public Future wraps() { return wraps; }

  Future wrapped()
  {
    if (wraps == null) throw UnsupportedErr.make("Future missing wraps");
    return wraps;
  }

  private Future wraps;
}

File: src/concurrent/dotnet/Future.cs
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   30 Mar 09  Andy Frank  Creation
//

using System.Collections;
using System.Runtime.CompilerServices;
using System.Threading;

namespace Fan.Sys
{
  /// <summary>
  /// Future is used to manage the entire lifecycle of each
  /// message send to an Actor.  An actor's queue is a linked
  /// list of messages.
  /// </summary>
  public sealed class Future : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    internal Future(object msg)
    {
      this.m_msg   = msg;
      this.m_state = PENDING;
    }

  //////////////////////////////////////////////////////////////////////////
  // Obj
  //////////////////////////////////////////////////////////////////////////

    public override Type @typeof()
    {
      if (m_type == null) m_type = Type.find("concurrent::Future");
      return m_type;
    }
    private static Type m_type;

  //////////////////////////////////////////////////////////////////////////
  // Future
  //////////////////////////////////////////////////////////////////////////

    public bool isDone()
    {
      return (m_state & DONE) != 0;
    }

    public bool isCancelled()
    {
      return m_state == DONE_CANCEL;
    }

    public object get() { return get(null); }
    public object get(Duration timeout)
    {
      object r = null;
      try
      {
        lock (this)
        {
          // wait until we enter a done state, the only notifies
          // on this object should be from cancel, set, or err
          if (timeout == null)
          {
            // wait forever until done
            while ((m_state & DONE) == 0) Monitor.Wait(this);
          }
          else
          {
            // if not done, then wait with timeout and then
            // if still not done throw a timeout exception
            if ((m_state & DONE) == 0)
            {
              Monitor.Wait(this, (int)timeout.millis());
              if ((m_state & DONE) == 0) throw TimeoutErr.make("Future.get timed out").val;
            }
          }

          // if canceled throw CancelErr
          if (m_state == DONE_CANCEL)
            throw CancelledErr.make("message canceled").val;

          // if error was raised, raise it to caller
          if (m_state == DONE_ERR)
            throw ((Err)m_result).rebase();

          // assign result to local variable for return
          r = m_result;
        }
      }
      catch (ThreadInterruptedException e)
      {
        throw InterruptedErr.make(e).val;
      }

      // ensure immutable or safe copy
      return Sys.safe(r);
    }

    public void cancel()
    {
      ArrayList wd;
      lock (this)
      {
        if ((m_state & DONE) == 0) m_state = DONE_CANCEL;
        m_msg = m_result = null;  // allow gc
        Monitor.PulseAll(this);
        wd = whenDone; whenDone = null;
      }
      sendWhenDone(wd);
    }

    internal void set(object r)
    {
      r = Sys.safe(r);
      ArrayList wd;
      lock (this)
      {
        m_state = DONE_OK;
        m_result = r;
        Monitor.PulseAll(this);
        wd = whenDone; whenDone = null;
      }
      sendWhenDone(wd);
    }

    internal void err(Err e)
    {
      ArrayList wd;
      lock (this)
      {
        m_state = DONE_ERR;
        m_result = e;
        Monitor.PulseAll(this);
        wd = whenDone; whenDone = null;
      }
      sendWhenDone(wd);
    }

  //////////////////////////////////////////////////////////////////////////
  // When Done
  //////////////////////////////////////////////////////////////////////////

    internal void sendWhenDone(Actor a, Future f)
    {
      // if already done, then set immediate flag
      // otherwise add to our when done list
      bool immediate = false;
      lock (this)
      {
        if (isDone()) immediate = true;
        else
        {
          if (whenDone == null) whenDone = new ArrayList();
          whenDone.Add(new WhenDone(a, f));
        }
      }

      // if immediate we are already done so enqueue immediately
      if (immediate)
      {
        try { a._enqueue(f, false); }
        catch (System.Exception e) { Err.dumpStack(e); }
      }
    }

    internal static void sendWhenDone(ArrayList list)
    {
      if (list == null) return;
      for (int i=0; i<list.Count; ++i)
      {
        WhenDone wd = (WhenDone)list[i];
        try { wd.actor._enqueue(wd.future, false); }
        catch (System.Exception e) { Err.dumpStack(e); }
      }
    }

    internal class WhenDone
    {
      public WhenDone(Actor a, Future f) { actor = a; future = f; }
      public Actor actor;
      public Future future;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    const int PENDING     = 0x00;
    const int DONE        = 0x0f;
    const int DONE_CANCEL = 0x1f;
    const int DONE_OK     = 0x2f;
    const int DONE_ERR    = 0x4f;

    internal object m_msg;         // message send to Actor
    internal Future m_next;        // linked list in Actor
    private volatile int m_state;  // processing state of message
    private object m_result;       // result or exception of processing
    private ArrayList whenDone;    // list of messages to deliver when done

  }
}